File: backend/routers/tasks.py
from datetime import datetime, timezone
from typing import List, Optional

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import func

from database import db_session
from db_models import Bucket, Project, Task
from models import TaskCreate, TaskMove, TaskResponse, TaskUpdate
from storage import (
    delete_task_file,
    generate_next_id,
    read_task_file,
    write_task_file,
)

router = APIRouter(prefix="/projects/{project_id}/tasks", tags=["Tasks"])


@router.get("", response_model=List[TaskResponse])
def get_tasks(
    project_id: str,
    bucket: Optional[str] = Query(None, description="Filter by bucket (column) name"),
    buckets: Optional[str] = Query(None, description="Comma-separated list of bucket names to include"),
    tag: Optional[str] = Query(None, description="Filter by tag name"),
    tags: Optional[str] = Query(None, description="Comma-separated list of tags to filter by"),
    tag_mode: Optional[str] = Query("any", description="Tag matching mode: 'any' (default) or 'all'"),
    exclude_bucket: Optional[str] = Query(None, description="Exclude tasks from this bucket name"),
    priorities: Optional[str] = Query(None, description="Comma-separated list of priorities (low, medium, high, urgent, none)"),
    search: Optional[str] = Query(None, description="Case-insensitive search in task title and body"),
    due_before: Optional[str] = Query(None, description="Filter tasks due on or before this date (YYYY-MM-DD)"),
    due_after: Optional[str] = Query(None, description="Filter tasks due on or after this date (YYYY-MM-DD)"),
    has_due_date: Optional[bool] = Query(None, description="Filter tasks that have a due date (True) or don't (False)"),
):
    with db_session() as session:
        # Verify project exists
        project = session.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Project '{project_id}' not found.",
            )

        query = session.query(Task).filter(Task.project_id == project_id)

        # 1. Bucket filtering
        if buckets:
            bucket_list = [b.strip() for b in buckets.split(",") if b.strip()]
            if bucket_list:
                query = query.filter(Task.bucket.in_(bucket_list))
        elif bucket:
            query = query.filter(Task.bucket == bucket)
        elif exclude_bucket:
            query = query.filter(Task.bucket != exclude_bucket)

        # 2. Priority filtering
        if priorities:
            priority_list = [p.strip().lower() for p in priorities.split(",") if p.strip()]
            if priority_list:
                conds = []
                if "none" in priority_list:
                    conds.append(Task.priority.is_(None))
                    conds.append(Task.priority == "")
                    priority_list = [p for p in priority_list if p != "none"]
                if priority_list:
                    conds.append(Task.priority.in_(priority_list))

                from sqlalchemy import or_

                query = query.filter(or_(*conds))

        # 3. Search filtering
        if search:
            search_pattern = f"%{search}%"
            from sqlalchemy import or_

            query = query.filter(or_(Task.title.ilike(search_pattern), Task.body.ilike(search_pattern)))

        # 4. Due Date filtering
        if has_due_date is not None:
            if has_due_date:
                query = query.filter(Task.due_date.isnot(None)).filter(Task.due_date != "")
            else:
                from sqlalchemy import or_

                query = query.filter(or_(Task.due_date.is_(None), Task.due_date == ""))

        if due_before:
            query = query.filter(Task.due_date.isnot(None)).filter(Task.due_date != "").filter(Task.due_date <= due_before)

        if due_after:
            query = query.filter(Task.due_date.isnot(None)).filter(Task.due_date != "").filter(Task.due_date >= due_after)

        query = query.order_by(Task.position.asc())
        tasks = query.all()

        # 5. Tag filtering
        filter_tags = []
        if tags:
            filter_tags = [t.strip().lower() for t in tags.split(",") if t.strip()]
        elif tag:
            filter_tags = [tag.strip().lower()]

        if filter_tags:
            mode = tag_mode.strip().lower() if tag_mode else "any"
            if mode == "all":
                tasks = [t for t in tasks if all(any(tg.lower() == ft for tg in t.tags) for ft in filter_tags)]
            else:
                tasks = [t for t in tasks if any(any(tg.lower() == ft for tg in t.tags) for ft in filter_tags)]

        return [
            TaskResponse(
                id=t.id,
                project_id=t.project_id,
                title=t.title,
                bucket=t.bucket,
                position=t.position,
                tags=t.tags,
                body=t.body,
                due_date=t.due_date,
                priority=t.priority,
                color=t.color,
                created_at=t.created_at,
                updated_at=t.updated_at,
            )
            for t in tasks
        ]


@router.get("/{task_id}", response_model=TaskResponse)
def get_task(project_id: str, task_id: str):
    task_data = read_task_file(task_id)
    if not task_data or task_data["project_id"] != project_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with ID {task_id} not found in project '{project_id}'",
        )
    return TaskResponse(**task_data)


@router.post("", response_model=TaskResponse, status_code=status.HTTP_201_CREATED)
def create_task(project_id: str, task: TaskCreate):
    with db_session() as session:
        # Verify project exists
        project = session.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Project '{project_id}' not found.",
            )

        # Verify bucket exists for this project
        db_bucket = session.query(Bucket).filter(Bucket.project_id == project_id, Bucket.name == task.bucket).first()
        if not db_bucket:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Bucket '{task.bucket}' does not exist in project '{project_id}'.",
            )

        now_str = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
        new_id = generate_next_id()

        # Calculate position: min position in current bucket - 1000.0 (or default to 1000.0)
        min_pos = session.query(func.min(Task.position)).filter(Task.project_id == project_id, Task.bucket == task.bucket).scalar()
        new_position = 1000.0 if min_pos is None else float(min_pos) - 1000.0

        task_tags = [t.lower() for t in task.tags]
        task_data = {
            "id": new_id,
            "project_id": project_id,
            "title": task.title,
            "bucket": task.bucket,
            "position": new_position,
            "tags": task_tags,
            "body": task.body,
            "due_date": task.due_date,
            "priority": task.priority,
            "color": task.color,
            "created_at": now_str,
            "updated_at": now_str,
        }

        # Write to Markdown file
        filename = write_task_file(new_id, task_data)

        # Insert into SQLite index
        db_task = Task(
            id=new_id,
            project_id=project_id,
            title=task.title,
            bucket=task.bucket,
            position=new_position,
            tags=task_tags,
            filename=filename,
            body=task.body,
            due_date=task.due_date,
            priority=task.priority,
            color=task.color,
            created_at=now_str,
            updated_at=now_str,
        )
        session.add(db_task)
        session.flush()

        return TaskResponse(**task_data)


@router.put("/{task_id}", response_model=TaskResponse)
def update_task(project_id: str, task_id: str, task_update: TaskUpdate):
    # Fetch existing task from Markdown
    existing = read_task_file(task_id)
    if not existing or existing["project_id"] != project_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with ID {task_id} not found in project '{project_id}'",
        )

    now_str = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")

    # Merge updates
    updated_title = task_update.title if task_update.title is not None else existing["title"]
    updated_bucket = task_update.bucket if task_update.bucket is not None else existing["bucket"]
    updated_position = task_update.position if task_update.position is not None else existing["position"]
    updated_tags = [t.lower() for t in task_update.tags] if task_update.tags is not None else [t.lower() for t in existing["tags"]]
    updated_body = task_update.body if task_update.body is not None else existing["body"]
    # due_date and priority support explicit null to clear the field
    updated_due_date = task_update.due_date if "due_date" in task_update.model_fields_set else existing.get("due_date")
    updated_priority = task_update.priority if "priority" in task_update.model_fields_set else existing.get("priority")
    updated_color = task_update.color if "color" in task_update.model_fields_set else existing.get("color")

    updated_data = {
        "id": task_id,
        "project_id": project_id,
        "title": updated_title,
        "bucket": updated_bucket,
        "position": updated_position,
        "tags": updated_tags,
        "body": updated_body,
        "due_date": updated_due_date,
        "priority": updated_priority,
        "color": updated_color,
        "created_at": existing["created_at"],
        "updated_at": now_str,
    }

    with db_session() as session:
        # Verify new bucket exists if changed
        if task_update.bucket is not None:
            bucket_exists = session.query(Bucket).filter(Bucket.project_id == project_id, Bucket.name == updated_bucket).first()
            if not bucket_exists:
                if updated_bucket == "done":
                    max_pos = session.query(func.max(Bucket.position)).filter(Bucket.project_id == project_id).scalar()
                    new_position = 1000.0 if max_pos is None else float(max_pos) + 1000.0

                    db_bucket = Bucket(
                        project_id=project_id,
                        name="done",
                        title="Done",
                        subtitle="",
                        position=new_position,
                    )
                    session.add(db_bucket)
                    session.flush()

                    all_buckets = session.query(Bucket).filter(Bucket.project_id == project_id).order_by(Bucket.position.asc()).all()
                    from storage import write_buckets_file

                    all_buckets_dict = []
                    for b in all_buckets:
                        all_buckets_dict.append(
                            {
                                "name": b.name,
                                "title": b.title,
                                "subtitle": b.subtitle,
                                "position": b.position,
                                "color": b.color,
                                "layout": b.layout,
                                "max_tasks": b.max_tasks,
                                "is_default": b.is_default,
                            }
                        )
                    write_buckets_file(project_id, all_buckets_dict)
                else:
                    raise HTTPException(
                        status_code=status.HTTP_400_BAD_REQUEST,
                        detail=f"Bucket '{updated_bucket}' does not exist in project '{project_id}'.",
                    )

        # Write to Markdown file
        new_filename = write_task_file(task_id, updated_data)

        # Update SQLite index
        db_task = session.query(Task).filter(Task.id == task_id, Task.project_id == project_id).first()
        if db_task:
            db_task.title = updated_title
            db_task.bucket = updated_bucket
            db_task.position = updated_position
            db_task.tags = updated_tags
            db_task.filename = new_filename
            db_task.body = updated_body
            db_task.due_date = updated_due_date
            db_task.priority = updated_priority
            db_task.color = updated_color
            db_task.updated_at = now_str
            session.flush()

    return TaskResponse(**updated_data)


@router.patch("/{task_id}/move", response_model=TaskResponse)
def move_task(project_id: str, task_id: str, task_move: TaskMove):
    # Fetch existing task from Markdown
    existing = read_task_file(task_id)
    if not existing or existing["project_id"] != project_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with ID {task_id} not found in project '{project_id}'",
        )

    now_str = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")

    updated_data = {
        **existing,
        "bucket": task_move.bucket,
        "position": task_move.position,
        "updated_at": now_str,
    }

    with db_session() as session:
        # Verify target bucket exists for this project
        bucket_exists = session.query(Bucket).filter(Bucket.project_id == project_id, Bucket.name == task_move.bucket).first()
        if not bucket_exists:
            if task_move.bucket == "done":
                max_pos = session.query(func.max(Bucket.position)).filter(Bucket.project_id == project_id).scalar()
                new_position = 1000.0 if max_pos is None else float(max_pos) + 1000.0

                db_bucket = Bucket(
                    project_id=project_id,
                    name="done",
                    title="Done",
                    subtitle="",
                    position=new_position,
                )
                session.add(db_bucket)
                session.flush()

                all_buckets = session.query(Bucket).filter(Bucket.project_id == project_id).order_by(Bucket.position.asc()).all()
                from storage import write_buckets_file

                all_buckets_dict = []
                for b in all_buckets:
                    all_buckets_dict.append(
                        {
                            "name": b.name,
                            "title": b.title,
                            "subtitle": b.subtitle,
                            "position": b.position,
                            "color": b.color,
                            "layout": b.layout,
                            "max_tasks": b.max_tasks,
                            "is_default": b.is_default,
                        }
                    )
                write_buckets_file(project_id, all_buckets_dict)
            else:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Bucket '{task_move.bucket}' does not exist in project '{project_id}'.",
                )

        # Write to Markdown file
        new_filename = write_task_file(task_id, updated_data)

        # Update SQLite index
        db_task = session.query(Task).filter(Task.id == task_id, Task.project_id == project_id).first()
        if db_task:
            db_task.bucket = task_move.bucket
            db_task.position = task_move.position
            db_task.filename = new_filename
            db_task.updated_at = now_str
            session.flush()

    return TaskResponse(**updated_data)


@router.delete("/{task_id}")
def delete_task(project_id: str, task_id: str):
    # Verify task exists in the project
    existing = read_task_file(task_id)
    if not existing or existing["project_id"] != project_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Task with ID {task_id} not found in project '{project_id}'",
        )

    # Delete from filesystem
    delete_task_file(task_id)

    # Delete from SQLite
    with db_session() as session:
        db_task = session.query(Task).filter(Task.id == task_id, Task.project_id == project_id).first()
        if db_task:
            session.delete(db_task)
            session.flush()

    return {"status": "success", "detail": f"Task {task_id} deleted"}

File: internal/handlers/tasks.go
package handlers

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"jotter/backend/internal/db"
	"jotter/backend/internal/models"
	"jotter/backend/internal/storage"
)

func RegisterTaskRoutes(r chi.Router, tasksDir string) {
	r.Get("/projects/{project_id}/tasks", func(w http.ResponseWriter, r *http.Request) {
		projectID := chi.URLParam(r, "project_id")

		// Verify project exists
		var dummy string
		err := db.DB.QueryRow("SELECT id FROM projects WHERE id = ?", projectID).Scan(&dummy)
		if err == sql.ErrNoRows {
			SendError(w, http.StatusNotFound, fmt.Sprintf("Project '%s' not found.", projectID))
			return
		} else if err != nil {
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}

		query := "SELECT id, project_id, title, bucket, position, tags, body, due_date, priority, color, created_at, updated_at FROM tasks WHERE project_id = ?"
		var args []interface{}
		args = append(args, projectID)

		bucket := r.URL.Query().Get("bucket")
		if bucket != "" {
			query += " AND bucket = ?"
			args = append(args, bucket)
		}

		excludeBucket := r.URL.Query().Get("exclude_bucket")
		if excludeBucket != "" {
			query += " AND bucket != ?"
			args = append(args, excludeBucket)
		}

		// Priority filtering
		priorities := r.URL.Query().Get("priorities")
		if priorities != "" {
			pList := strings.Split(priorities, ",")
			var activePriorities []string
			hasNone := false
			for _, p := range pList {
				p = strings.TrimSpace(p)
				if strings.ToLower(p) == "none" {
					hasNone = true
				} else if p != "" {
					activePriorities = append(activePriorities, strings.ToLower(p))
				}
			}

			if hasNone || len(activePriorities) > 0 {
				query += " AND ("
				subConditions := []string{}
				if hasNone {
					subConditions = append(subConditions, "priority IS NULL", "priority = ''")
				}
				if len(activePriorities) > 0 {
					var placeholders []string
					for _, p := range activePriorities {
						placeholders = append(placeholders, "?")
						args = append(args, p)
					}
					subConditions = append(subConditions, fmt.Sprintf("priority IN (%s)", strings.Join(placeholders, ",")))
				}
				query += strings.Join(subConditions, " OR ")
				query += ")"
			}
		}

		query += " ORDER BY position ASC"

		rows, err := db.DB.Query(query, args...)
		if err != nil {
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}
		defer rows.Close()

		var taskList []models.TaskResponse
		for rows.Next() {
			var t models.TaskResponse
			var tagsJSON string
			var dueDate, priority, color sql.NullString

			err := rows.Scan(&t.ID, &t.ProjectID, &t.Title, &t.Bucket, &t.Position, &tagsJSON, &t.Body, &dueDate, &priority, &color, &t.CreatedAt, &t.UpdatedAt)
			if err != nil {
				SendError(w, http.StatusInternalServerError, err.Error())
				return
			}

			_ = json.Unmarshal([]byte(tagsJSON), &t.Tags)
			if dueDate.Valid {
				t.DueDate = &dueDate.String
			}
			if priority.Valid {
				t.Priority = &priority.String
			}
			if color.Valid {
				t.Color = &color.String
			}

			taskList = append(taskList, t)
		}

		if taskList == nil {
			taskList = []models.TaskResponse{}
		}

		SendJSON(w, http.StatusOK, taskList)
	})

	r.Post("/projects/{project_id}/tasks", func(w http.ResponseWriter, r *http.Request) {
		projectID := chi.URLParam(r, "project_id")

		if r.Body == nil {
			log.Printf("ERROR: POST /tasks - Request body is nil")
			SendError(w, http.StatusBadRequest, "Request body is missing")
			return
		}

		var req models.TaskCreate
		bodyBytes, _ := io.ReadAll(r.Body)
		if len(bodyBytes) == 0 {
			log.Printf("ERROR: POST /tasks - Request body is empty")
			SendError(w, http.StatusBadRequest, "Request body is empty")
			return
		}
		// Restore body for decoding
		r.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))

		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			log.Printf("ERROR: POST /tasks - Failed to decode JSON: %v", err)
			SendError(w, http.StatusBadRequest, "Invalid request payload")
			return
		}

		// Verify project exists
		var dummy string
		err := db.DB.QueryRow("SELECT id FROM projects WHERE id = ?", projectID).Scan(&dummy)
		if err == sql.ErrNoRows {
			SendError(w, http.StatusNotFound, fmt.Sprintf("Project '%s' not found.", projectID))
			return
		} else if err != nil {
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}

		// Verify bucket exists for this project
		err = db.DB.QueryRow("SELECT name FROM buckets WHERE project_id = ? AND name = ?", projectID, req.Bucket).Scan(&dummy)
		if err == sql.ErrNoRows {
			SendError(w, http.StatusBadRequest, fmt.Sprintf("Bucket '%s' does not exist in project '%s'.", req.Bucket, projectID))
			return
		} else if err != nil {
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}

		nowStr := time.Now().UTC().Format(time.RFC3339Nano)
		nowStr = strings.Replace(nowStr, "+00:00", "Z", 1)
		newID := storage.GenerateULID()

		// Calculate position: min position - 1000.0
		var minPos sql.NullFloat64
		_ = db.DB.QueryRow("SELECT MIN(position) FROM tasks WHERE project_id = ? AND bucket = ?", projectID, req.Bucket).Scan(&minPos)
		newPosition := 1000.0
		if minPos.Valid {
			newPosition = minPos.Float64 - 1000.0
		}

		var tags []string
		for _, t := range req.Tags {
			tags = append(tags, strings.ToLower(t))
		}

		taskMap := map[string]interface{}{
			"id":         newID,
			"project_id": projectID,
			"title":      req.Title,
			"bucket":     req.Bucket,
			"position":   newPosition,
			"tags":       tags,
			"body":       req.Body,
			"due_date":   req.DueDate,
			"priority":   req.Priority,
			"color":      req.Color,
			"created_at": nowStr,
			"updated_at": nowStr,
		}

		filename, errWrite := storage.WriteTaskFile(tasksDir, newID, taskMap)
		if errWrite != nil {
			SendError(w, http.StatusInternalServerError, "Failed to write task file")
			return
		}

		// Write to DB
		tagsJSON, _ := json.Marshal(tags)
		var dbDueDate, dbPriority, dbColor sql.NullString
		if req.DueDate != nil {
			dbDueDate = sql.NullString{String: *req.DueDate, Valid: true}
		}
		if req.Priority != nil {
			dbPriority = sql.NullString{String: *req.Priority, Valid: true}
		}
		if req.Color != nil {
			dbColor = sql.NullString{String: *req.Color, Valid: true}
		}

		_, err = db.DB.Exec("INSERT INTO tasks (id, project_id, title, bucket, position, tags, filename, body, due_date, priority, color, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
			newID, projectID, req.Title, req.Bucket, newPosition, string(tagsJSON), filename, req.Body, dbDueDate, dbPriority, dbColor, nowStr, nowStr)
		if err != nil {
			// Clean up file if DB fails
			_ = storage.DeleteTaskFile(tasksDir, newID)
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}

		res := models.TaskResponse{
			ID:        newID,
			ProjectID: projectID,
			Title:     req.Title,
			Bucket:    req.Bucket,
			Position:  newPosition,
			Tags:      tags,
			Body:      req.Body,
			DueDate:   req.DueDate,
			Priority:  req.Priority,
			Color:     req.Color,
			CreatedAt: nowStr,
			UpdatedAt: nowStr,
		}
		SendJSON(w, http.StatusCreated, res)
	})

	r.Route("/projects/{project_id}/tasks/{task_id}", func(r chi.Router) {
		r.Get("/", func(w http.ResponseWriter, r *http.Request) {
			projectID := chi.URLParam(r, "project_id")
			taskID := chi.URLParam(r, "task_id")

			taskData, err := storage.ReadTaskFile(tasksDir, taskID)
			if err != nil || taskData.ProjectID != projectID {
				SendError(w, http.StatusNotFound, fmt.Sprintf("Task with ID %s not found in project '%s'", taskID, projectID))
				return
			}

			SendJSON(w, http.StatusOK, taskData)
		})

		r.Delete("/", func(w http.ResponseWriter, r *http.Request) {
			projectID := chi.URLParam(r, "project_id")
			taskID := chi.URLParam(r, "task_id")

			existing, errRead := storage.ReadTaskFile(tasksDir, taskID)
			if errRead != nil || existing.ProjectID != projectID {
				SendError(w, http.StatusNotFound, fmt.Sprintf("Task with ID %s not found in project '%s'", taskID, projectID))
				return
			}

			deleted := storage.DeleteTaskFile(tasksDir, taskID)
			if !deleted {
				SendError(w, http.StatusInternalServerError, "Failed to delete task file")
				return
			}

			_, err := db.DB.Exec("DELETE FROM tasks WHERE id = ? AND project_id = ?", taskID, projectID)
			if err != nil {
				SendError(w, http.StatusInternalServerError, err.Error())
				return
			}

			SendJSON(w, http.StatusOK, map[string]string{
				"status": "success",
				"detail": fmt.Sprintf("Task %s deleted", taskID),
			})
		})

		updateHandler := func(w http.ResponseWriter, r *http.Request) {
			projectID := chi.URLParam(r, "project_id")
			taskID := chi.URLParam(r, "task_id")

			bodyBytes, err := io.ReadAll(r.Body)
			if err != nil {
				SendError(w, http.StatusBadRequest, "Failed to read request body")
				return
			}

			var raw map[string]interface{}
			if err := json.Unmarshal(bodyBytes, &raw); err != nil {
				SendError(w, http.StatusBadRequest, "Invalid JSON payload")
				return
			}

			var req models.TaskUpdate
			if err := json.Unmarshal(bodyBytes, &req); err != nil {
				SendError(w, http.StatusBadRequest, "Invalid request payload")
				return
			}

			existing, errRead := storage.ReadTaskFile(tasksDir, taskID)
			if errRead != nil || existing.ProjectID != projectID {
				SendError(w, http.StatusNotFound, fmt.Sprintf("Task with ID %s not found in project '%s'", taskID, projectID))
				return
			}

			nowStr := time.Now().UTC().Format(time.RFC3339Nano)
			nowStr = strings.Replace(nowStr, "+00:00", "Z", 1)

			updatedTitle := existing.Title
			if _, ok := raw["title"]; ok && req.Title != nil {
				updatedTitle = *req.Title
			}

			updatedBucket := existing.Bucket
			if _, ok := raw["bucket"]; ok && req.Bucket != nil {
				updatedBucket = *req.Bucket
			}

			updatedPosition := existing.Position
			if _, ok := raw["position"]; ok && req.Position != nil {
				updatedPosition = *req.Position
			}

			var updatedTags []string
			if _, ok := raw["tags"]; ok && req.Tags != nil {
				for _, tg := range *req.Tags {
					updatedTags = append(updatedTags, strings.ToLower(tg))
				}
			} else {
				updatedTags = existing.Tags
			}

			updatedBody := existing.Body
			if _, ok := raw["body"]; ok && req.Body != nil {
				updatedBody = *req.Body
			}

			updatedDueDate := existing.DueDate
			if _, ok := raw["due_date"]; ok {
				updatedDueDate = req.DueDate
			}

			updatedPriority := existing.Priority
			if _, ok := raw["priority"]; ok {
				updatedPriority = req.Priority
			}

			updatedColor := existing.Color
			if _, ok := raw["color"]; ok {
				updatedColor = req.Color
			}

			tx, err := db.DB.Begin()
			if err != nil {
				SendError(w, http.StatusInternalServerError, err.Error())
				return
			}
			defer func() { _ = tx.Rollback() }()

			if _, ok := raw["bucket"]; ok && req.Bucket != nil && *req.Bucket != existing.Bucket {
				var bDummy string
				errB := tx.QueryRow("SELECT name FROM buckets WHERE project_id = ? AND name = ?", projectID, updatedBucket).Scan(&bDummy)
				if errB == sql.ErrNoRows {
					if updatedBucket == "done" {
						var maxPos sql.NullFloat64
						_ = tx.QueryRow("SELECT MAX(position) FROM buckets WHERE project_id = ?", projectID).Scan(&maxPos)
						newPosition := 1000.0
						if maxPos.Valid {
							newPosition = maxPos.Float64 + 1000.0
						}
						_, err = tx.Exec("INSERT INTO buckets (project_id, name, title, subtitle, position, color, layout, max_tasks, is_default) VALUES (?, 'done', 'Done', '', ?, NULL, 'list', NULL, 0)",
							projectID, newPosition)
						if err != nil {
							SendError(w, http.StatusInternalServerError, err.Error())
							return
						}
					} else {
						SendError(w, http.StatusBadRequest, fmt.Sprintf("Bucket '%s' does not exist in project '%s'.", updatedBucket, projectID))
						return
					}
				} else if errB != nil {
					SendError(w, http.StatusInternalServerError, errB.Error())
					return
				}
			}

			taskMap := map[string]interface{}{
				"project_id": projectID,
				"title":      updatedTitle,
				"bucket":     updatedBucket,
				"position":   updatedPosition,
				"tags":       updatedTags,
				"body":       updatedBody,
				"due_date":   updatedDueDate,
				"priority":   updatedPriority,
				"color":      updatedColor,
				"created_at": existing.CreatedAt,
				"updated_at": nowStr,
			}

			filename, errWrite := storage.WriteTaskFile(tasksDir, taskID, taskMap)
			if errWrite != nil {
				SendError(w, http.StatusInternalServerError, "Failed to write task file")
				return
			}

			tagsJSON, _ := json.Marshal(updatedTags)
			var dbDueDate, dbPriority, dbColor sql.NullString
			if updatedDueDate != nil {
				dbDueDate = sql.NullString{String: *updatedDueDate, Valid: true}
			}
			if updatedPriority != nil {
				dbPriority = sql.NullString{String: *updatedPriority, Valid: true}
			}
			if updatedColor != nil {
				dbColor = sql.NullString{String: *updatedColor, Valid: true}
			}

			_, err = tx.Exec("UPDATE tasks SET title = ?, bucket = ?, position = ?, tags = ?, filename = ?, body = ?, due_date = ?, priority = ?, color = ?, updated_at = ? WHERE id = ? AND project_id = ?",
				updatedTitle, updatedBucket, updatedPosition, string(tagsJSON), filename, updatedBody, dbDueDate, dbPriority, dbColor, nowStr, taskID, projectID)
			if err != nil {
				SendError(w, http.StatusInternalServerError, err.Error())
				return
			}

			if err := tx.Commit(); err != nil {
				SendError(w, http.StatusInternalServerError, err.Error())
				return
			}

			if _, ok := raw["bucket"]; ok && updatedBucket == "done" {
				_ = syncBucketsFile(tasksDir, projectID)
			}

			res := models.TaskResponse{
				ID:        taskID,
				ProjectID: projectID,
				Title:     updatedTitle,
				Bucket:    updatedBucket,
				Position:  updatedPosition,
				Tags:      updatedTags,
				Body:      updatedBody,
				DueDate:   updatedDueDate,
				Priority:  updatedPriority,
				Color:     updatedColor,
				CreatedAt: existing.CreatedAt,
				UpdatedAt: nowStr,
			}
			SendJSON(w, http.StatusOK, res)
		}

		r.Put("/", updateHandler)
		r.Patch("/", updateHandler)
	})

	r.Patch("/projects/{project_id}/tasks/{task_id}/move", func(w http.ResponseWriter, r *http.Request) {
		projectID := chi.URLParam(r, "project_id")
		taskID := chi.URLParam(r, "task_id")

		var req models.TaskMove
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			SendError(w, http.StatusBadRequest, "Invalid request payload")
			return
		}

		existing, errRead := storage.ReadTaskFile(tasksDir, taskID)
		if errRead != nil || existing.ProjectID != projectID {
			SendError(w, http.StatusNotFound, fmt.Sprintf("Task with ID %s not found in project '%s'", taskID, projectID))
			return
		}

		nowStr := time.Now().UTC().Format(time.RFC3339Nano)
		nowStr = strings.Replace(nowStr, "+00:00", "Z", 1)

		tx, err := db.DB.Begin()
		if err != nil {
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}
		defer func() { _ = tx.Rollback() }()

		// Verify target bucket exists for this project
		var bDummy string
		errB := tx.QueryRow("SELECT name FROM buckets WHERE project_id = ? AND name = ?", projectID, req.Bucket).Scan(&bDummy)
		if errB == sql.ErrNoRows {
			if req.Bucket == "done" {
				var maxPos sql.NullFloat64
				_ = tx.QueryRow("SELECT MAX(position) FROM buckets WHERE project_id = ?", projectID).Scan(&maxPos)
				newPosition := 1000.0
				if maxPos.Valid {
					newPosition = maxPos.Float64 + 1000.0
				}

				_, err = tx.Exec("INSERT INTO buckets (project_id, name, title, subtitle, position, color, layout, max_tasks, is_default) VALUES (?, 'done', 'Done', '', ?, NULL, 'list', NULL, 0)",
					projectID, newPosition)
				if err != nil {
					SendError(w, http.StatusInternalServerError, err.Error())
					return
				}
			} else {
				SendError(w, http.StatusBadRequest, fmt.Sprintf("Bucket '%s' does not exist in project '%s'.", req.Bucket, projectID))
				return
			}
		} else if errB != nil {
			SendError(w, http.StatusInternalServerError, errB.Error())
			return
		}

		taskMap := map[string]interface{}{
			"project_id": projectID,
			"title":      existing.Title,
			"bucket":     req.Bucket,
			"position":   req.Position,
			"tags":       existing.Tags,
			"body":       existing.Body,
			"due_date":   existing.DueDate,
			"priority":   existing.Priority,
			"color":      existing.Color,
			"created_at": existing.CreatedAt,
			"updated_at": nowStr,
		}

		filename, errWrite := storage.WriteTaskFile(tasksDir, taskID, taskMap)
		if errWrite != nil {
			SendError(w, http.StatusInternalServerError, "Failed to write task file")
			return
		}

		_, err = tx.Exec("UPDATE tasks SET bucket = ?, position = ?, filename = ?, updated_at = ? WHERE id = ? AND project_id = ?",
			req.Bucket, req.Position, filename, nowStr, taskID, projectID)
		if err != nil {
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}

		if err := tx.Commit(); err != nil {
			SendError(w, http.StatusInternalServerError, err.Error())
			return
		}

		if req.Bucket == "done" {
			_ = syncBucketsFile(tasksDir, projectID)
		}

		res := models.TaskResponse{
			ID:        taskID,
			ProjectID: projectID,
			Title:     existing.Title,
			Bucket:    req.Bucket,
			Position:  req.Position,
			Tags:      existing.Tags,
			Body:      existing.Body,
			DueDate:   existing.DueDate,
			Priority:  existing.Priority,
			Color:     existing.Color,
			CreatedAt: existing.CreatedAt,
			UpdatedAt: nowStr,
		}
		SendJSON(w, http.StatusOK, res)
	})
}
